### ISSUE TITLE
torch.distributions.kl_divergence gives different gradients than manual implementation

### ISSUE THREAD

**Issue by chentau** [pytorch/pytorch#30090]
## 🐛 Bug

For some reason, the built in `torch.distributions.kl_divergence` is giving me different gradients wrt the parameters of the distributions, compared to when I manually implement the kl divergence.
As an example, I took the kl divergence of the categorical distribution - I haven't tested with any other distributions yet.

## To Reproduce

```python
import torch
import torch.distributions as dist

# built in kl divergence
a = dist.Dirichlet(torch.ones(5)).sample()
a.requires_grad = True
d1 = dist.Categorical(a)
d2 = dist.Categorical(torch.ones(5))

kl_1 = dist.kl_divergence(d1, d2)
print("kl 1: {}".format(kl_1))
kl_1.backward()
print(a.grad)

# manual implementation
a.grad.data = torch.zeros(5)
kl_2 = torch.sum(a * (torch.log(a) - np.log(1. / 5)))
print("kl 2: {}".format(kl_2))
kl_2.backward()
print(a.grad)
```

The two numbers should match up, but they do not, at least for me.

I also recognize that there is a very high likelihood it is something wrong with my code itself, and not the implementation of the kl divergence, so if that turns out to be the case, it would be great if you let me know where the error is.

## Environment

 PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.10
GCC version: (Ubuntu 8.3.0-6ubuntu1~18.10.1) 8.3.0
CMake version: version 3.12.1

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] numpydoc==0.9.1
[pip] torch==1.3.0
[pip] torchvision==0.4.0a0
[conda] _pytorch_select           0.2                       gpu_0  
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.0.2            py36h7b6447c_0  
[conda] mkl_fft                   1.0.12           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.2.0           cuda100py36h938c94c_0  
[conda] tensorflow                1.13.1          mkl_py36h27d456a_0  
[conda] tensorflow-base           1.13.1          mkl_py36h7ce6ba3_0  
[conda] torch                     1.3.0                    pypi_0    pypi
[conda] torchvision               0.4.0           cuda100py36hecfc37a_0

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by nikitaved** [comment]
Hi! Shouldn't d2 be `d2 = dist.Categorical(torch.ones(5) / 5)`?

**Comment by nikitaved** [comment]
Interesting. I confirm, I have the same issue. The manual gradient is correct, it matches
`torch.log(a) - np.log(1./5) + 1.`.

**Comment by SsnL** [comment]
Hmm so in this case actually both are correct. PyTorch is computing the prob with `a / a.sum()`, and thus get you a different gradient (with the same function value):
https://github.com/pytorch/pytorch/blob/c2e576e74b73bdb1eb9db30268883b8efa884de8/torch/distributions/categorical.py#L50

**Comment by fritzo** [comment]
One difference is that `dist.Categorical` normalizes its input, whereas your version allows gradient updates to change the norm.

**Comment by chentau** [comment]
If I understand correctly then, the built in and manually implemented gradients are proportional to each other, correct?  In that case, is one preferable to the other?  I would imagine that the built in normalization of the categorical distribution would be useful for prevent gradient explosion.

**Comment by nikitaved** [comment]
> If I understand correctly then, the built in and manually implemented gradients are proportional to each other, correct? In that case, is one preferable to the other? I would imagine that the built in normalization of the categorical distribution would be useful for preventing gradient explosion.

No, they are not proportional. So, given the normalization, i.e. that `a` sums up to 1, the true gradient in the first case, if my calculus is correct, becomes:
`(1-a)*(torch.log(a) - np.log(1./5) + 1)`,
while in the second case:
`torch.log(a) - np.log(1./5) + 1`.
The second gradient matches the one received from the backward method, while the first one does not. So, unless there is a mistake in my calculations, it looks like a bug.

**Comment by SsnL** [comment]
It’s really not a bug. The two functions match exactly on the prob simplex,
but differ otherwise. So they have different gradients.

**Comment by nikitaved** [comment]
I agree, that they have different gradients, but in the first case, when `a` is passed to the Categorical, the gradient does not match the one that I derived by hand, which is
`(1-a)*(torch.log(a) - np.log(1./5) + 1)` in my case. Is there a mistake in my derivation?

**Comment by vincentqb** [comment]
> Hmm so in this case actually both are correct. PyTorch is computing the prob with `a / a.sum()`, and thus get you a different gradient (with the same function value):
> https://github.com/pytorch/pytorch/blob/c2e576e74b73bdb1eb9db30268883b8efa884de8/torch/distributions/categorical.py#L50

I second that `(torch.log(d1) - np.log(1./5) + 1)` would be the gradient with respect to `d1` (if we could record it) and not `a`, at least because of what you are saying @SsnL.  It'd be nice to figure out the analytical version of what pytorch is doing. I was expecting what @nikitaved said too, `(1-a)*(torch.log(a) - np.log(1./5) + 1)` but that doesn't match either :) I any case, I would not call this a bug yet.

**Comment by fritzo** [comment]
I believe PyTorch computes a gradient that avoids numerical drift in logits space. Usually categorical distributions are parametrized by log-probabilities, named "logits" in the codebase. When computing gradients wrt logits, the PyTorch gradient sums to zero:
```py
import numpy as np
import torch
import torch.distributions as dist
from torch.autograd import grad

# built in kl divergence
logits = dist.Dirichlet(torch.ones(5)).sample().log()
logits.requires_grad = True
probs = logits.exp()
d1 = dist.Categorical(probs)
d2 = dist.Categorical(torch.ones(5) / 5)

kl_1 = dist.kl_divergence(d1, d2)
print("kl 1: {}".format(kl_1))
feasible_grad = grad(kl_1, [logits], retain_graph=True)[0]
print(feasible_grad)
print(feasible_grad.sum())
# tensor(-8.3819e-09)

# manual implementation
kl_2 = torch.sum(probs * (torch.log(probs) - np.log(1. / 5)))
print("kl 2: {}".format(kl_2))
infeasible_grad = grad(kl_2, [logits])[0]
print(infeasible_grad)
print(infeasible_grad.sum())
# tensor(1.3398)
```
I guess it would reasonable to instead avoid drift in `probs` space by ensuring that `grad(kl_1, [probs])` sums to zero.

**Comment by nikitaved** [comment]
@fritzo , the code supports both probabilities and logits. And if probabilities are supplied, the logits are the transformation of the probabilities and the other way around. Anyway, no matter what is transformation of what, we are interested in the gradient wrt `a`. And the output that torch produces does not agree with the analytic solution, even though we consider the normalization process done in the Categorical constructor. So, either there is some other hidden transformation is going on somewhere, or it is a bug. The thing is, I did some debugging, and the only additional transformation is clamping values to be in [epsilon, 1-epsilon]. And, given that the sample values are in the this interval, it become essentially an identity transformation in most of the cases.

**Comment by nikitaved** [comment]
@SsnL, @fritzo, @vincentqb, sum is indeed doing something weird, check this out:
```python
torch.manual_seed(5)

x = dist.Dirichlet(torch.ones(5)).sample()
x.requires_grad = True
print(f"vector x is: {x}")
obj = torch.sum(x / torch.sum(x))
print(f"obj: {obj}")
obj.backward()
print(f"Experiment grad: {x.grad}")

Output:
vector x is: tensor([0.0546, 0.2794, 0.0830, 0.0952, 0.4878], requires_grad=True)
obj: 1.0
Experiment grad: tensor([0., 0., 0., 0., 0.])
```
The gradient should be `1-x`, not `0`.

**Comment by nikitaved** [comment]
@chentau @fritzo @vincentqb the analytic gradient for the normalized case
is `(torch.log(a) + 1) - torch.sum(a * (torch.log(a) + 1))`, not `(1-a)*(torch.log(a) - np.log(1./5) + 1)`. The correct formula matches the gradient from backward.

My mistake, I am sorry. Closing the issue...
